File: src/matcher.ts
export function pathMatchesGlob(path: string, pattern: string): boolean {
  path = path.replace(/\\/g, "/");
  pattern = pattern.replace(/\\/g, "/");

  let regex = "^";
  let i = 0;
  while (i < pattern.length) {
    const c = pattern[i];
    if (c === "*") {
      if (i + 1 < pattern.length && pattern[i + 1] === "*") {
        regex += ".*";
        i += 2;
        if (i < pattern.length && pattern[i] === "/") {
          regex += "/?";
          i++;
        }
      } else {
        regex += "[^/]*";
        i++;
      }
    } else if (c === "?") {
      regex += "[^/]";
      i++;
    } else {
      regex += c.replace(/[.+^${}()|[\]\\]/g, "\\$&");
      i++;
    }
  }
  regex += "$";

  return new RegExp(regex).test(path);
}

export function anyFileMatches(files: string[], patterns: string[]): boolean {
  return files.some((f) => patterns.some((p) => pathMatchesGlob(f, p)));
}

File: autodoc_hooks/matcher.py
import re


def path_matches_glob(path: str, pattern: str) -> bool:
    path = path.replace("\\", "/")
    pattern = pattern.replace("\\", "/")

    parts: list[str] = []
    i = 0
    while i < len(pattern):
        c = pattern[i]
        if c == "*":
            if i + 1 < len(pattern) and pattern[i + 1] == "*":
                parts.append(".*")
                i += 2
                # consume the trailing slash so foo/** matches foo/bar not foo//bar
                if i < len(pattern) and pattern[i] == "/":
                    parts.append("/?")
                    i += 1
            else:
                parts.append("[^/]*")
                i += 1
        elif c == "?":
            parts.append("[^/]")
            i += 1
        else:
            parts.append(re.escape(c))
            i += 1

    return bool(re.match("^" + "".join(parts) + "$", path))


def any_file_matches(files: list[str], patterns: list[str]) -> bool:
    return any(path_matches_glob(f, p) for f in files for p in patterns)
